### PR TITLE
[Workflow] Update workflow storage interface

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

The current APIs requires "storage" as an input, this is inconvenient and hard to control. We decide to use global storage initialized by `workflow.init()`

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/workflow/api.py [hunk 1]
```diff
@@ -8,16 +9,42 @@
...
+def init(storage: "Optional[Union[str, Storage]]" = None) -> None:
+    """Initialize workflow.
+
+    Args:
+        storage: The external storage URL or a custom storage class. If not
+            specified, ``/tmp/ray/workflow_data`` will be used.
+    """
+    if not ray.is_initialized():
+        logger.info("Initializing Ray...")
+        ray.init(address="auto")
```
Comment (ericl) [inline python/ray/experimental/workflow/api.py:31]: Don't need this anymore btw (Ray does auto init by itself)

File: python/ray/experimental/workflow/storage/__init__.py [hunk 2]
```diff
@@ -52,17 +51,8 @@ def create_storage(storage_url: str) -> Storage:
...
-            # We should use get_temp_dir_path, but for ray client, we don't
-            # have this one. We need a flag to tell whether it's a client
-            # or a driver to use the right dir.
-            # For now, just use /tmp/ray/workflow_data
-            logger.warning(
-                "Using default local dir: `/tmp/ray/workflow_data`. "
-                "This should only be used for testing purposes.")
-            storage_url = "file:///tmp/ray/workflow_data"
-        _global_storage = create_storage(storage_url)
+        raise ValueError("There is no global storage. Maybe you forget to "
```
Comment (ericl) [inline python/ray/experimental/workflow/storage/__init__.py:54]:
```suggestion
        raise ValueError("`workflow.init()` must be run prior to using the workflows API. "
```

File: python/ray/experimental/workflow/api.py [hunk 3]
```diff
@@ -8,16 +9,42 @@
...
+    workflow_access.init_management_actor()
```
Comment (iycheng) [inline python/ray/experimental/workflow/api.py:32]: should we pass the storage url to the actor?

File: python/ray/experimental/workflow/workflow_access.py [hunk 4]
```diff
@@ -203,6 +214,9 @@ def get_or_create_management_actor() -> "ActorHandle":
         actor = ray.get_actor(MANAGEMENT_ACTOR_NAME)
     except ValueError:
         # the actor does not exist
+        logger.warning("Cannot access workflow manager. It could because "
+                       "the workflow manager exited unexpectedly. A new "
+                       "workflow manager is being created.")
```
Comment (iycheng) [inline python/ray/experimental/workflow/workflow_access.py:219]: I feel these two functions are actually the same, right?

**Review by fishbone — COMMENTED** [review]
Thanks for the updating! For the coordinator actor here, we probably want to pass URL to that actor.
When doing workflow.init, we need to check the storage URL matches or not. With this, the coordinator actor will always hold the store internally.

Right now we don't want to complicate the cases here and only want to support one cluster one storage.

File: python/ray/experimental/workflow/workflow_access.py (hunk 4 above)
Comment (suquark) [reply python/ray/experimental/workflow/workflow_access.py:219]: basically the same, but one gives you a warning and returns the actor, one not.

File: python/ray/experimental/workflow/api.py (hunk 1 above)
Comment (suquark) [reply python/ray/experimental/workflow/api.py:31]: oh, that an interesting fact I did not know!

File: python/ray/experimental/workflow/api.py (hunk 3 above)
Comment (suquark) [reply python/ray/experimental/workflow/api.py:32]: then should we have something like `workflow.shutdown` to terminate the actor? and how to know the actor uses the same storage? should I check it whenever I use the actor?

**Comment by suquark** [comment]
@iycheng  I feel confused about passing storage URL to that actor:

1. Should we check the workflow is using the same storage with the actor?
2. If there will be multiple drivers, who decides which storage the actor will use?

File: python/ray/experimental/workflow/workflow_access.py (hunk 4 above)
Comment (iycheng) [reply python/ray/experimental/workflow/workflow_access.py:219]: It's ok but just doesn't look nice. Since there are not a lot of lines, maybe it's fine.

**Comment by iycheng** [comment]
> @iycheng I feel confused about passing storage URL to that actor:
> 
> 1. Should we check the workflow is using the same storage with the actor?

We should check the workflow is using the same storage in `workflow.init`.

> 2. If there will be multiple drivers, who decides which storage the actor will use?

We don't allow multiple storages for the same actor. The actor will store the URL when init. If the workflow.init(url) is different from the one actor stored, throw an exception.

--------

The thing here is that the requirement for multiple storage support is not very clear and when developing if written in this way, we need to keep in mind that the actor can handle multiple storages.

If later we decide to support this, I'd prefer hash(storage_url) as the actor name, so that, we are sure it'll connect to the right actor. For now, let's just make things simple here.

File: python/ray/experimental/workflow/api.py [hunk 5]
```diff
@@ -8,16 +9,52 @@
...
+        storage = "file:///tmp/ray/workflow_data"
+    if isinstance(storage, str):
+        storage = storage_base.create_storage(storage)
+    elif not isinstance(storage, Storage):
+        raise TypeError("'storage' should be None, str, or Storage type.")
+    try:
+        _storage = storage_base.get_global_storage()
+    except RuntimeError:
+        pass
+    else:
```
Comment (iycheng) [inline python/ray/experimental/workflow/api.py:47]: Add some "empty" lines to separate the blocks for easy understanding. It's a little bit hard to distinguish the blocks here.

File: python/ray/experimental/workflow/workflow_access.py [hunk 6]
```diff
@@ -193,16 +200,38 @@ def report_success(self, workflow_id: str) -> None:
         self._workflow_outputs.pop(workflow_id, None)
 
 
+def init_management_actor() -> None:
+    """Initialize WorkflowManagementActor"""
+    store = storage.get_global_storage()
+    try:
+        workflow_manager = ray.get_actor(MANAGEMENT_ACTOR_NAME)
+        is_valid = workflow_manager.validate_storage.remote(store.storage_url)
```
Comment (iycheng) [inline python/ray/experimental/workflow/workflow_access.py:208]: I think here it's wrong, right? You need ray.get to get the internal value. Btw, this logic should belong to init_management_actor, not the actual actor.

How about this
```
if ray.get(workflow_manager.storage_url.remote()) != storage_url:
    raise RuntimeError(...)
```

**Review by fishbone — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Pull request merged**